File: CSharp/TLE&ACByCpp/EscapeALargeMaze.cpp
//
// Created by user3301 on 5/6/2019.
//
#include <vector>
#include <unordered_set>
#include <queue>
#include <sstream>
using namespace std;
class EscapeALargeMaze {
public:
    bool isEscapePossible(vector<vector<int>>& blocked, vector<int>& source, vector<int>& target) {
        if(blocked.size() == 0) return true;
        unordered_set<string> blockers;
        for (int i = 0;
             i < blocked.size(); ++i) {
            stringstream ss;
            ss<<blocked[i][0]<<","<<blocked[i][1];
            blockers.emplace(ss.str());
        }
        unordered_set<string> seen;
        return bfs(blockers, source, target, seen) && bfs(blockers, target, source, seen);
    }

private:
    bool bfs(unordered_set<string>& blockers, vector<int>& source, vector<int>& target, unordered_set<string>& seen) {
        if(target[0] < 0 || target[1]< 0 || target[0]>=1e6 || target[1]>= 1e6) return false;

        vector<vector<int>> dirs{{1,0},{0,1},{-1,0}, {0,-1}};

        queue<vector<int>> q;

        q.emplace(source);

        while(q.size()!=0) {
            int size = q.size();

            for (int i = 0;
                 i < size; ++i) {
                auto point = q.front();
                q.pop();
                if(point[0]==target[0] && point[1]==target[1]) return true;

                for(auto dir:dirs) {
                    int x = point[0] + dir[0];
                    int y = point[1] + dir[1];
                    stringstream ss;
                    ss<<x<<","<<y;
                    string str = ss.str();
                    if(x >=0 && x < 1e6 && y >=0 && y < 1e6 && blockers.find(str)==blockers.end() && seen.find(str)==seen.end()) {
                        if(target[0]==x && target[1] == y) return true;
                        vector<int> temp{x,y};
                        q.emplace(temp);
                        seen.emplace(str);
                        if(seen.size() >=2e4) return true;
                    }
                }
            }
        }
        return false;
    }
};

File: CSharp/leetcodesln/EscapeALargeMaze.cs
using System.Collections.Generic;

namespace leetcodesln
{
    public class EscapeALargeMaze
    {

        /// <summary>
        ///  This is a correct solution but TLE, it has been proved using c++
        /// </summary>
        /// <param name="blocked"></param>
        /// <param name="source"></param>
        /// <param name="target"></param>
        /// <returns></returns>
        public bool IsEscapePossible(int[][] blocked, int[] source, int[] target)
        {
            if (blocked.Length == 0) return true;
            var blockers = new HashSet<string>();
            foreach (var b in blocked)
            {
                blockers.Add(b[0] + "," + b[1]);
            }
            var seen = new HashSet<string>();
            return IsEscapePossibleBFS(blockers, source, target, ref seen) && IsEscapePossibleBFS(blockers, target, source, ref seen);
        }

        private bool IsEscapePossibleBFS(HashSet<string> blockers, int[] source, int[] target, ref HashSet<string> seen)
        {
            if (target[0] < 0 || target[1] < 0 || target[0] >= 1e6 || target[1] >= 1e6) return false;

            var dirs = new int[4][];
            dirs[0] = new int[] { 1, 0 };
            dirs[1] = new int[] { 0, 1 };
            dirs[2] = new int[] { -1, 0 };
            dirs[3] = new int[] { 0, -1 };


            var q = new Queue<int[]>();


            q.Enqueue(source);

            while (q.Count != 0)
            {
                int size = q.Count;

                for (int i = 0; i < size; i++)
                {
                    var point = q.Dequeue();
                    if (point[0] == target[0] && point[1] == target[1]) return true;

                    foreach (var dir in dirs)
                    {
                        int x = point[0] + dir[0];
                        int y = point[1] + dir[1];
                        var str = x + "," + y;
                        if (x >= 0 && x < 1e6 && y >= 0 && y < 1e6 && !blockers.Contains(str) && !seen.Contains(str))
                        {
                            if (target[0] == x && target[1] == y) return true;
                            q.Enqueue(new int[] { x, y });
                            seen.Add(str);
                            if (seen.Count == 2e4) return true;
                        }

                    }
                }
            }
            return false;
        }
    }
}
